pkg/ucp/store/inmemory/client.go

```diff
@@ -0,0 +1,262 @@
+	}
+	parsed, err := resources.Parse(id)
+	if err != nil {
+		return nil, &store.ErrInvalid{Message: "invalid argument. 'id' must be a valid resource id"}
+	}
+	if parsed.IsEmpty() {
+		return nil, &store.ErrInvalid{Message: "invalid argument. 'id' must not be empty"}
+	}
+	if parsed.IsResourceCollection() || parsed.IsScopeCollection() {
+		return nil, &store.ErrInvalid{Message: "invalid argument. 'id' must refer to a named resource, not a collection"}
+	}
+
+	normalized, err := storeutil.NormalizeResourceID(parsed)
+	if err != nil {
+		return nil, err
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	entry, ok := c.resources[strings.ToLower(normalized.String())]
+	if !ok {
+		return nil, &store.ErrNotFound{ID: id}
+	}
+
+	return &entry.obj, nil
+}
+
+// Delete implements store.StorageClient.
+func (c *Client) Delete(ctx context.Context, id string, options ...store.DeleteOptions) error {
+	if ctx == nil {
+		return &store.ErrInvalid{Message: "invalid argument. 'ctx' is required"}
+	}
+	parsed, err := resources.Parse(id)
+	if err != nil {
+		return &store.ErrInvalid{Message: "invalid argument. 'id' must be a valid resource id"}
+	}
+	if parsed.IsEmpty() {
+		return &store.ErrInvalid{Message: "invalid argument. 'id' must not be empty"}
+	}
+	if parsed.IsResourceCollection() || parsed.IsScopeCollection() {
+		return &store.ErrInvalid{Message: "invalid argument. 'id' must refer to a named resource, not a collection"}
+	}
+
+	normalized, err := storeutil.NormalizeResourceID(parsed)
+	if err != nil {
+		return err
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	config := store.NewDeleteConfig(options...)
+
+	entry, ok := c.resources[strings.ToLower(normalized.String())]
+	if !ok && config.ETag != "" {
+		return &store.ErrConcurrency{}
+	} else if !ok {
+		return &store.ErrNotFound{ID: id}
+	} else if config.ETag != "" && config.ETag != entry.obj.ETag {
+		return &store.ErrConcurrency{}
+	}
+
+	delete(c.resources, strings.ToLower(normalized.String()))
+
+	return nil
+}
+
+// Query implements store.StorageClient.
+func (c *Client) Query(ctx context.Context, query store.Query, options ...store.QueryOptions) (*store.ObjectQueryResult, error) {
+	if ctx == nil {
+		return nil, &store.ErrInvalid{Message: "invalid argument. 'ctx' is required"}
+	}
+
+	err := query.Validate()
+	if err != nil {
+		return nil, &store.ErrInvalid{Message: fmt.Sprintf("invalid argument. Query is invalid: %s", err.Error())}
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	result := &store.ObjectQueryResult{}
+	for _, entry := range c.resources {
+		// Check root scope.
+		if query.ScopeRecursive && !strings.HasPrefix(entry.rootScope, storeutil.NormalizePart(query.RootScope)) {
+			continue
+		} else if !query.ScopeRecursive && entry.rootScope != storeutil.NormalizePart(query.RootScope) {
+			continue
+		}
+
+		// Check resource type.
+		resourceType, err := storeutil.NormalizeResourceType(query.ResourceType)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
We can do this outside the for loop, right?

Reply:
We could. I'm not worried about optimizing the behavior of the in-memory store unless we have a reason to.